### ISSUE TITLE
shell: use of killpg(2) to terminate processes may break cleanup for some applications

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-core#3445]
This came up from a discussion with @FrankD412 on slack.

When a job is canceled, a `SIGTERM` is sent to all job shell processes which each then forward that signal to all tasks with `flux_subprocess_kill(3)`. By default I believe, `flux_subprocess_kill` uses `killpg(2)` to send the signal to the entire process group to ensure as much cleanup as possible (similar to a Unix shell forwarding SIGINT to the current process group which owns the terminal).

This is causing problems for @FrankD412, however, because their main process spaws worker subprocesses which unexpectedly receive SIGTERM on `flux job cancel`, interrupting a clean exit which requires that the main process instruct the workers to clean up. In this case, it would be better if the initial `SIGTERM` were sent _only_ to the main task, and the subsequent `SIGKILL` (after grace period) was sent to the process group.

Unfortunately, this would cause problems in the common case. That is, if a task creates children but _doesn't_ block SIGTERM, the main task would terminate on `flux job cancel` and the children would be left around. This is specifically the case that the use of `killpg(2)` is trying to handle.

However, maybe a shell option could be provided to change this behavior, e.g. `-o no-killpg` or similar could disable the use of `killpg(2)` in general by the shell could be used in cases where like this. The option would be ignored, of course, when the signal is `SIGKILL`.

Note that there are also other mechanisms than `flux job cancel` (which, in general, would be expected to kill jobs immediately) to tell a job is should gracefully exit. One idea would be for jobs to subscribe to exceptions and then generate a "job shutdown" exception with nonzero severity. This would allow the job to gracefully shutdown without the use of signals, but requires more work than a signal handler, so may not be preferred by application developers.

There may be other ideas for alternate non-fatal ways to trigger job cleanup as well.

**Comment by dongahn** [comment]
@FrankD412's email description was that those subprocesses are hung when the parent process is killed. @FrankD412: are the subprocesses also install SIGTERM and ignore it?

**Comment by FrankD412** [comment]
@dongahn -- Sorry for not getting back earlier. In tests on Lassen, I have been testing using individual handlers for each sub-process which simply log that they saw the `SIGTERM` and then just pass. That seems to have worked and is now resulting in an actual exception we need to figure out after all process join. The interesting bit is that we have not encountered that hang on Summit using the older version of the code (where each process does not have an explicit `SIGTERM` handler). On Summit, we see that the main process has caught the signal, and that the process all have attempted to join. We're not sure if we've simply been lucky or if something is fundamentally different, but there is definitely a difference between Lassen and Summit somewhere.

@grondo -- This might be a differing preference/opinion or inaccurate understanding on my part, but it is my understanding that `SIGINT` is more of a cease operation and relinquish control of the prompt immediately signal (and therefore is sent to the process group). I'll say that the below is meant to explain my mental model, not necessarily a plea for you to change your mind, after all you all are the owners of the project.

From my perspective, treating a running job with the same precedence as a user job running on the terminal is a bit of a mismatch in my mind. In my mental model, I treat the scheduler and cluster as a "computer with its own OS and scheduling policies" that when I submit something to it, it's a service or non-interactive job that resides on the system. With that model in mind, I don't view it as reasonable to ask a whole process group to cease immediately as if it were on my terminal; instead I view it more as sending a `kill <pid>` which defaults to a `SIGTERM`. To my understanding, that is usually sent to the highest level parent process (in this case the flux process that manages the running process, which then forwards to the user's job). It is then the user's code that is responsible for clean up in the form of forwarding the signal to subprocesses, joining its processes, etc -- much the same way that my local machine's OS doesn't evict a process from running, it simply asks the process to gracefully exit from the top level via a `SIGTERM`.

I'll caveat this with the following: I don't write MPI programs, so my mental model may be incorrect and asking all MPI subtasks to terminate may be reasonable. My mental model is very service oriented such that I would trust a user to write their software such that they can handle a graceful exit. It is only when they linger past the grace period that a hard kill can be sent. A decent compromise might be to send a "polite" `SIGTERM` to the root process, wait for a few seconds and then send a "more forceful" `SIGTERM` to the group, followed by a hard eviction `SIGKILL`. Again, this is with the mental model that it's the host process' responsibility to clean up its process tree (and the place of the scheduler to take the lightest action possible first) before taking heavier actions. Additionally, I don't think SLURM or LSF send a `SIGTERM` in the same fashion as a users would call for a `SIGINT` -- so maybe what I'm experiencing is a mismatch in expectations between different schedulers (I've not tested this explicitly, but it's from what I've witnessed in passing working with users so could be inaccurate). Again, I may be incorrect so please take this with a grain of salt if I'm out in left field.

**Comment by FrankD412** [comment]
@dongahn @grondo -- After some more debug, I was able to reproduce the issue I was seeing before on a login node (Lassen). I think we can rule that out as a Flux issue; but so far I'm suspecting it's an issue with child processes being killed out from a parent process in Python (so far, from limited information I've been able to find, it appears is **might** be a Python bug). Just wanted to pass along my findings, thanks!
